src/scripting/lua_unit.cpp

```diff
@@ -461,7 +461,13 @@ static int impl_unit_set(lua_State *L)
 	modify_vector_string_attrib("extra_recruit", u.set_recruits(value));
 	modify_vector_string_attrib("advances_to", u.set_advances_to(value));
 	if(strcmp(m, "alignment") == 0) {
-		u.set_alignment(lua_check<UNIT_ALIGNMENT>(L, 3));
+		std::string str = lua_check<std::string>(L, 3);
```

Comment:
It'd certainly be nice for `lua_check` to have an enum overload that does all this for you.

Reply:
I think it wouldn't be hard to support something like `lua_check<enum_definition_list>(L, idx)`?

Reply:
What would `enum_definition_list` be here?

Reply:
```
template<typename T>
std::enable_if_t<std::is_base_of_v<enum_base, T>, std::optional<typename T::type>>
lua_check(lua_State *L, int n)
{
	std::string str = lua_check_impl::lua_check<std::string>(L, n);
	std::optional<typename T::type> val = T::get_enum(str);
	if(!val) {
		luaL_argerror(L, n, ("cannot convert " + str + " to enum.").c_str());
	}
	return val;
}
```
is the closest I've been able to get, but this doesn't work with just
`lua_check<unit_alignments>(L, 3).value_or(unit_alignments::type::NEUTRAL)`
since that wants to return an instance of `unit_alignments`. What does work is:
`lua_check_impl::lua_check<unit_alignments>(L, 3).value_or(unit_alignments::type::NEUTRAL)`

Reply:
Huh? Doesn't `lua_check<unit_alignments>(L, 3)` work with that definition?

Reply:
It gives the error: `cannot convert ‘std::decay_t<string_enums::enum_base<unit_alignments_defines> >’ {aka ‘string_enums::enum_base<unit_alignments_defines>’} to ‘unit_alignments_defines::type’`

Reply:
Okay yeah, that's because of the `lua_check` wrapper in the global namespace that then calls the `lua_check_impl` version…

You could add another overload of the wrapper too, I guess… (or instead)

Reply:
So like:
```
template<typename T>
std::optional<typename T::type> lua_enum_check(lua_State *L, int n)
{
	//remove possible const& to make life easier for the impl namespace.
	return lua_check_impl::lua_check<T>(L, n);
}
```
?

Reply:
Um… well, sure, I guess that works.

Reply:
Actually, I wonder if it would make sense to use `luaL_checkoption` here since that's specifically intended for enums? It won't return an `std::optional` though, but that's probably preferred since `check` functions are supposed to raise a Lua error if the argument is invalid.

In other words, you can delete the implementation function and just write this:

```lua
template<typename T>
typename T::type lua_enum_check(lua_State *L, int n)
{
	return T::get_enum(luaL_checkoption(L, n, nullptr, T::values));
}
```

That seems correct at least; I haven't tested it. It won't return at all if the value is invalid. It does rely on the enum constants being contiguous (ie, no explicit constant assignments), but I think non-contiguous enums would already be a problem in the system, right?

Reply:
Right.

Reply:
Ah, I just noticed that [the documentation](https://www.lua.org/manual/5.4/manual.html#luaL_checkoption) states the _array_ must be null-terminated, so some kind of trickery might be needed to guarantee that's the case.

Reply:
Though it looks like `luaL_checkoption`'s final argument is `const char* const*` rather than `const std::array<const char*, 4>`.

Reply:
Well yes, but the latter is trivially convertible to the former (`values.data()`). The only tricky part is the null-terminator.

I'm not sure if consecutive static variables are guaranteed to be contiguous in memory, so my personal choice for solving this would probably be a wrapper type that inserts an extra "ghost" element, something like this:

```cpp
template<size_t n>
class enum_array : public std::array<const char*, n> {
  using std::array<const char*, n>::array;
  const char* null_terminator = nullptr;
};
```

This would mean that when `luaL_checkoption` tries to read off the end of the array, it would find the `null_terminator` there and stop. (This is admittedly assuming that there's no padding; but I don't think padding would be inserted between pointers anyway.) It's less specialized than built-in `std::array` because this trick only really works with pointers.

Maybe there are other solutions. I'm not sure.

One thing I personally wouldn't do is to just make the array one longer and manually add a nullptr to the end, because that would mess up your `get_string` that's assuming all the pointers are valid. But you could still make it work as long as you account for this.

Reply:
I think I'd prefer to leave it as it is currently to be honest, given it's only used in once place. If in the future it's used in more places and ends up needing a different implementation, then I'd rather address that then.

Reply:
At the very least I think it makes sense for it to return `T::type` instead of `std::optional<T::type>`.

Reply:
It would need to have a default value returned if the conversion fails, in that case.

Reply:
No, because it _doesn't return at all_ if the conversion fails (it raises a Lua error instead).

Reply:
Ah, my bad. Updated.